cluster-autoscaler/simulator/drain.go

```diff
@@ -68,25 +69,28 @@ func NewNodeDeleteOptions(opts config.AutoscalingOptions) NodeDeleteOptions {
 func GetPodsToMove(nodeInfo *schedulerframework.NodeInfo, deleteOptions NodeDeleteOptions, listers kube_util.ListerRegistry,
 	pdbs []*policyv1.PodDisruptionBudget, timestamp time.Time) (pods []*apiv1.Pod, daemonSetPods []*apiv1.Pod, blockingPod *drain.BlockingPod, err error) {
 	var drainPods, drainDs []*apiv1.Pod
+	drainabilityRules := deleteOptions.DrainabilityRules
+	if drainabilityRules == nil {
```

Comment:
I'm not a fan of this defaulting, seems a bit like pushing the initialisation into the implementation. We already have this set in `NewNodeDeleteOptions`, which should be enough to make it always used. Additionally this can make testing more complex.

I get that we are now putting the core functionality into a parameter, which is a bit scary. But maybe we can do this better by splitting `DrainabilityRules` into something like `CoreDrainabilityRules` and `CustomDrainabilityRules`? This way we will ensure that the core rules are always executed, while also giving a way for the user to define their own rules if they so desire. This does not allow to override the core rules, but maybe that's for the better? User can still go around it if we execute the custom rules before core ones.

Reply:
The thing is that sometimes we may want a custom rule that is evaluated before the core rules, sometimes after them. And sometimes - instead of them. For example, today we have a bunch of flags that control drainability behavior. Rather than having rules that check the flag value every time, I was envisioning this as picking the set of rules once based on the flags. If there's immutable "core" part, we won't be able to do this.

> Additionally this can make testing more complex.

Actually, this defaulting here is meant to make tests work. There's a [bunch of them](https://github.com/search?q=repo%3Akubernetes%2Fautoscaler+NodeDeleteOptions+path%3A**_test.go&type=code) that don't use `NewNodeDeleteOptions` to create the `NodeDeleteOptions`:  Alternatively, I can go and fix all these tests instead, perhaps this is going to be cleaner than handling `nil` here that way?

Reply:
Ok, I see. Initially I thought that it would be still doable with just the "prefix rules", like youwould be able to override any default behaviour if you defined rules before. But the "UndefinedOutcome" cannot be passed to force end iteration over rules, so that wouldn't work :/

As a side note - maybe a drainabilityStatus which works as `UndefinedOutcome` (passes the pod to `GetPodsForDeletionOnNodeDrain`) *but* force ends iteration over rules would be useful?

Reply:
It would make sense if `GetPodsForDeletionOnNodeDrain` was here to stay, but the idea was to migrate all of it into drainability rules.